\documentclass[nojss]{jss}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% GENERAL HINTS: STYLE OF THIS DOCUMENT
%
% Use \pkg{} for package names.
% Use \proglang{} for classes.
% Use \code{} for R commands.
% Use `` and '' for opening and closing double quotes, respectively.
% Put a single sentence in a single line, i.e. start a new line for each new sentence.
% Use empty lines for indicating paragraphs.
% Put several empty lines before the start of a new section.
% Put several empty lines surrounding a figure caption.
% Put words such as "via" in italics.
% Put taxon names in italics (but not strain names).
% Put TODOs in \rcsmark or \rcscom tags and remove them once done.
% Be concise and avoid repetitions, as they cause a huge maintenance problem.
% Avoid the term "functionality". It had been used way too frequently.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Add Revision Control System (RCS) Markups
\usepackage{ulem}      % use this for sout
%\usepackage{fancyvrb}
\normalem              % set \emph to \textit again
% Additions (in blue)
\def\rcsadd#1{{\color{blue}#1}}
% Deletions (in red)
\def\rcsdel#1{{\color{red}\sout{#1}}}
% Yellow box (few words)
\def\rcsmark#1{\colorbox{yellow}{#1}}
%Yellow box (paragraph)
\def\rcscom#1{\noindent\newline\vspace*{0.5cm}\colorbox{yellow}{\parbox{\textwidth}{#1}}\vspace*{0.5cm}}

%% needed for CRAN checking
%\VignetteIndexEntry{Using opm}

% authors, first page
\author{Johannes Sikorski\\Leibniz Institute DSMZ  \And
        Lea A.I. Vaas\\Leibniz Institute DSMZ \AND
        Benjamin Hofner\\Universit\"{a}t \\Erlangen-N\"{u}rnberg \And
        Markus G\"{o}ker\\Leibniz Institute DSMZ}

% title, first page
\title{\pkg{opm}: An R Package for Analysing OmniLog\textregistered \ Phenotype MicroArray Data}

% authors, header on every 2nd page
\Plainauthor{J. Sikorski, L.A.I. Vaas, B. Hofner, M. G\"{o}ker}

% main title, first page
\Plaintitle{opm: An R package for analysing OmniLog Phenotype MicroArray Data}

% short title, header on every 2nd page
\Shorttitle{Phenotype MicroArray Data (\today)}

\Abstract{
here some abstrac-text.
This is the tutorial of \pkg{opm} in the version of \today{}.

}

\Keywords{Bootstrap, Cell Lines, \pkg{grofit}, Growth Curves, \pkg{lattice}, Metadata, Microbiology, Respiration Kinetics, Splines, \proglang{YAML}, \proglang{JSON}, \proglang{CSV}}

\Plainkeywords{bootstrap, cell lines, grofit, growth curves, lattice, metadata, microbiology, respiration kinetics, splines, YAML, JSON, CSV}

\Address{
  Markus G\"{o}ker\\
  Leibniz Institute DSMZ -- German Collection of Microorganisms and Cell Cultures\\
  Braunschweig\\
  \\
  Telephone: +49/531-2616-272\\
  Fax: +49/531-2616-237\\
  E-mail: \email{markus.goeker@dsmz.de}\\
  URL: \url{www.dsmz.de}\\
}


%% this must be included if Sweave is used (with % symbols):
%% need no \usepackage{Sweave.sty}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\begin{document}
\SweaveOpts{concordance=TRUE}
% NOTE: Rstudio might insert a line '\SweaveOpts{concordance=TRUE}' after '\begin{document}', which is removed again by the build script, as it causes warning messages during R CMD check.
% So please don't try to rescue that line, and don't rely on it being there.


%% so far adding a TOC proved unsuccessful -- apparently the JSS style doesn't allow for a TOC
%\tableofcontents
%\newpage


\section[Introduction]{Introduction}\label{introduction}

\subsection[]{}\label{}

After comparative analyses of Phenotype Microarray data, users might be interested in mapping the data on pathway maps for biological interpretation of higher-level systemic functions.
% erklären, wozu man die daten nutzen kann, wenn man grafik und gruppenvergleiche gemacht hat

% übersicht über vorhandene platten, informationen über abfrage von substratinformationen und suchen
% welche funktionalitäten überhaupt da sind -> grafik für workflow?

KEGG is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies.
Pathway Mapping
KEGG PATHWAY mapping is the process to map molecular datasets, especially large-scale datasets in genomics, transcriptomics, proteomics, and metabolomics, to the KEGG pathway maps for biological interpretaion of higher-level systemic functions. 

\subsection[]{}\label{}

\section[Acknowledgements]{Acknowledgements}

The help of Nora Buddruhs (DSMZ) and Anne Fiebig (DSMZ), who contributed a lot to the stored substrate information and to examples in earlier versions of this tutorial, is gratefully acknowledged.
We owe very much to Hans-Peter Klenk (DSMZ), who brought the OmniLog\textregistered \ instrument to the DSMZ and supported this project in numerous ways.
We thank Barry Bochner (BIOLOG Inc.), John Kirkish (BIOLOG Inc.), Andre Chouankam (BIOLOG Inc.), Jan Meier-Kolthoff (DSMZ), Pia W\"{u}st (DSMZ), Stefan Ehrentraut (DSMZ) and J\"{o}rn Petersen (DSMZ) for helpful advice, as well as Victoria Michael (DSMZ) for technical support.
The integration of missing OmniLog\textregistered \ substrates into ChEBI by the ChEBI staff is gratefully acknowledged.
We are also grateful to the maintainers of R-Forge and CRAN for providing the online resources used by this project.
This work was supported by the German Research Foundation (DFG) SFB/TRR 51 and by the Microme project within the Framework 7 programme of the European Commission, which is gratefully acknowledged.
JS gratefully acknowledges his support by DFG grant SI 1352/1-2.


\bibliography{opm}

\end{document}

